mlir/test/Dialect/Affine/affine-loop-interchange.mlir

```diff
@@ -0,0 +1,264 @@
+        affine.store %c0, %A[%i, %j] : memref<2048x2048xf64>
+      }
+    } else {
+      affine.for %j = 0 to 2048 {
+        affine.store %c1, %A[%i, %j] : memref<2048x2048xf64>
+      }
+    }
+  }
+  return
+}
+
+// -----
+
+// Test for interchange on imperfect nests.
+
+func @imperfect_nest(%A: memref<2048x2048xf64>) {
+  %c0 = constant 0.0 : f64
+  %c1 = constant 1.0 : f64
+  affine.for %i = 0 to 2048 {
+    affine.for %j = 0 to 1024 {
+      affine.store %c0, %A[%j, %i] : memref<2048x2048xf64>
+    }
+    affine.for %j = 1024 to 2048 {
+      affine.store %c1, %A[%j, %i] : memref<2048x2048xf64>
+    }
+  }
+  return
+}
+// CHECK:      for %{{.*}} = 0 to 1024 {
+// CHECK-NEXT:   for %{{.*}} = 0 to 2048 {
+// CHECK-NEXT:     affine.store %{{.*}}, %arg0[%arg1, %arg2]
+// CHECK-NEXT:   }
+// CHECK-NEXT: }
+// CHECK-NEXT: for %{{.*}} = 1024 to 2048 {
+// CHECK-NEXT:   for %{{.*}} = 0 to 2048 {
+// CHECK-NEXT:      affine.store %{{.*}}, %arg0[%arg1, %arg2]
+// CHECK-NEXT:   }
+// CHECK-NEXT: }
+
+
+// -----
+
+// Test for interchange on multilevel imperfect nests.
+
+func @multilevel_imperfect_nest(%A: memref<2048x2048xf64>) {
+  %c0 = constant 0.0 : f64
+  %c1 = constant 1.0 : f64
+  affine.for %i = 0 to 2048 {
+    affine.for %j = 0 to 1024 {
+      affine.store %c0, %A[%j, %i] : memref<2048x2048xf64>
+    }
+    affine.for %j = 1024 to 2048 {
+      affine.store %c1, %A[%j, %i] : memref<2048x2048xf64>
+    }
+    affine.for %j = 1024 to 2048 {
+      affine.for %k = 0 to 256 {
+        affine.store %c0, %A[%j, %k] : memref<2048x2048xf64>
+      }
+      affine.for %k = 256 to 512 {
+        affine.store %c1, %A[%k, %i] : memref<2048x2048xf64>
+      }
+    }
+  }
+  return
+}
+
+// -----
+
+func @multi_nest_seq(%A: memref<2048x2048xf64>, %B: memref<2048x2048xf64>, %C: memref<2048x2048xf64>) {
+  affine.for %i = 0 to 2048 {
+    affine.for %j = 0 to 2048 {
+      %v = affine.load %A[%j, %i] : memref<2048x2048xf64>
+      affine.store %v, %A[%j, %i] : memref<2048x2048xf64>
+      // CHECK:       affine.load %arg0[%arg3, %arg4] : memref<2048x2048xf64>
+      // CHECK-NEXT:  affine.store %{{.*}}, %arg0[%arg3, %arg4] : memref<2048x2048xf64>
+    }
+  }
+
+  affine.for %i = 0 to 2048 {
+    affine.for %j = 0 to 2048 {
+      affine.for %k = 0 to 2048 {
+        %a = affine.load %A[%i, %k] : memref<2048x2048xf64>
+        %b = affine.load %B[%k, %j] : memref<2048x2048xf64>
+        %ci = affine.load %C[%i, %j] : memref<2048x2048xf64>
+        %p = mulf %a, %b : f64
+        %co = addf %ci, %p : f64
+        affine.store %co, %C[%i, %j] : memref<2048x2048xf64>
+      }
+    }
+  }
+  // CHECK:      affine.load %arg0[%arg3, %arg4] : memref<2048x2048xf64>
+  // CHECK-NEXT: affine.load %arg1[%arg4, %arg5] : memref<2048x2048xf64>
+  // CHECK-NEXT: affine.load %arg2[%arg3, %arg5] : memref<2048x2048xf64>
+  // CHECK-NEXT: mulf
+  // CHECK-NEXT: addf
+  // CHECK-NEXT: affine.store %{{.*}}, %arg2[%arg3, %arg5] : memref<2048x2048xf64>
+  return
+}
```

Comment:
Missing test case where interchange is invalid due to dependences.

Reply:
Fixed. Thanks.